### DISCUSSION TITLE
Request body - overwrite arrays

### DISCUSSION THREAD

**Discussion by thim81** [apideck-libraries/portman#285] [Q&A] [answered]
The OpenAPI conversion seems to generate a request body with 2 items? How to overwrite certain elements in the array?

Example OpenAPI:

```yaml
openapi: 3.0.3
info:
  version: "1"
  title: 282-Workspace Desks
  contact: {}
paths:
  /workplaces/desks:
    post:
      operationId: CreateDesks
      summary: Create new desk
      description: Create a new desk not assigned to a specific user
      tags:
        - Desks
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/CreateDeskDetail"
        required: true
      responses:
        "201":
          description: "Created"
components:
  schemas:
    CreateDeskDetail:
      properties:
        data:
          type: array
          items:
            type: object
            properties:
              key_1:
                type: string
                example: abcdef
              key_2:
                type: string
                example: abcdef
              key_3:
                type: string
                example: abcdef
              key_4:
                type: string
                example: abcdef
              key_5:
                type: string
                example: abcdef
              key_6:
                type: string
                example: abcdef

```

Which results in Postman in: 
<img width="700" alt="2022-03-22 at 17 55 57@2x" src="https://user-images.githubusercontent.com/952446/159533853-bc8bba98-0a46-4078-8222-371e2d383d96.png">

How can I overwrite the `key_1`, `key_2`, ... values in the `data` array and only have 1 item in the array?

**Comment by thim81** [comment, marked as answer]
Portman uses the [openapi-to-postman](https://github.com/postmanlabs/openapi-to-postman) package to handle the conversion from OpenAPI to Postman.
One of the side effects the package has, is that it when there is OpenAPI property with type array, the generated data will have 2 elements which is valid according to the schema. See https://github.com/postmanlabs/openapi-to-postman/issues/410 for more details.

So there are a number of options to overcome this behaviour:


## 1. Define a maxItems for your array

If you'd like to define a custom number of elements in the generated array, you can mention the `minItems` and `maxItems` fields according to your requirements. i.e. following will generate only one element in the array.

```yaml
      requestBody:
        content:
          application/json:
            schema:
              description: The payload for sending a e-mail messages in bulk.
              type: array
              items:
                $ref: '#/components/schemas/itemObject'
              maxItems: 1
```

## 2. Use an `examples` reference in OpenAPI

By defining an actual example in your OpenAPI and not the example on single properties, the conversion will use the full example including a single array item.

```yaml
paths:
  '/email/v1/contacts':
    post:
      requestBody:
        content:
          application/json:
            schema:
              description: The payload for multiple contacts
              type: array
              items:
                $ref: '#/components/schemas/AllContacts'
components:
  schemas:
    AllContacts:
      type: array
      items:
        $ref: '#/components/schemas/ContactModel'
      example:
        - id: 1
          first_name: Sherlock
          last_name: Holmes

```

## 3. Use Portman `overwrites`

By overwriting certain items in the 1st element in the array and removing the 2nd element in the array, you will end-up with a single request body.

Example Portman configuration, based on the OpenAPI example:

```json
{
  "version": 1.0,
  "overwrites": [
    {
      "openApiOperation": "POST::/workplaces/desks",
      "overwriteRequestBody": [
        {
          "key": "data[0].key_1",
          "value": "XXXXXXXX",
          "overwrite": true
        },
        {
          "key": "data[0].key_2",
          "value": "XXXXXXXX",
          "overwrite": true
        },
        {
          "key": "data[0].key_3",
          "value": "XXXXXXXX",
          "overwrite": true
        },
        {
          "key": "data[0].key_4",
          "value": "XXXXXXXX",
          "overwrite": true
        },
        {
          "key": "data[0].key_5",
          "value": "XXXXXXXX",
          "overwrite": true
        },
        {
          "key": "data[0].key_6",
          "value": "XXXXXXXX",
          "overwrite": true
        },
        {
          "key": "data[1]",
          "remove": true
        }
      ]
    }
  ]
}
```

will result in a Postman request like this:
<img width="664" alt="2022-03-22 at 18 17 06@2x" src="https://user-images.githubusercontent.com/952446/159538146-75a5971e-bb5a-4dde-b74a-1a8aa65aea88.png">

**Comment by jdevosMagento** [comment]
I had tried this solution, but I hadn't removed the "value" variable...

I confirm that it works without 

Thank you!
